### ISSUE TITLE
elemental-operator installation sometimes fails

### ISSUE THREAD

**Issue by juadk** [rancher/elemental-operator#438]
In the CI, we can see that the elemental-operator (dev) installation fails sporadically with this message:

```
msg: "[helm upgrade --install elemental-operator oci://registry.opensuse.org/isv/rancher/elemental/dev/charts/rancher/elemental-operator-chart --namespace cattle-elemental-system --create-namespace] cmd, failed with the following error: Release \"elemental-operator\" does not exist.Installing it now.\nPulled: registry.opensuse.org/isv/rancher/elemental/dev/charts/rancher/elemental-operator-chart:1.2.2\nDigest: sha256:66f55a119c4ac7e3d5aed309b8510fd3288d4a027fef934b37a8e5b20b29f80e\nError: failed post-install: unable to build kubernetes object for deleting hook elemental-operator/templates/channel-hook.yaml: resource mapping not found for name: \"elemental-teal-channel\" namespace: \"fleet-default\" from \"\": no matches for kind \"ManagedOSVersionChannel\" in version \"elemental.cattle.io/v1beta1\"\nensure CRDs are installed first\n",
```

Failed job:
https://github.com/rancher/elemental/actions/runs/4887884892/jobs/8725003338

It also fails when the operator is uninstalled as well:
https://github.com/rancher/elemental/actions/runs/4889180147/jobs/8727492483

**Comment by davidcassany** [comment]
Feels like a race condition on the Helm's K8s client. CRDs are created but not searchable right after as the client pulls data from a cache. Seams to be a similar issue as reported here https://github.com/fluxcd/helm-controller/issues/655

**Comment by fgiudici** [comment]
Seems there are [best practices meant to avoid this issue](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#install-a-crd-declaration-before-using-the-resource).
We should either leverage the `crds` directory (would require to drop templating from our crds) or add a separate chart just for CRDs.

**Comment by juadk** [comment]
It happens less often but still there:
https://github.com/rancher/elemental/actions/runs/4954357027/jobs/8862787664

```
failed post-install: unable to build kubernetes object for post-install hook elemental-operator/templates/channel-hook.yaml: resource mapping not found for name: \"elemental-teal-channel\" namespace: \"fleet-default\" from \"\": no matches for kind \"ManagedOSVersionChannel\" in version \"elemental.cattle.io/v1beta1\"\nensure CRDs are installed first\n",
```

**Comment by fgiudici** [comment]
Recap of current status:
The issue is that we now have a CRD instance we load together with the CRDs definitions in the same chart (the default `ManagedOSVersionChannel`). This may lead to the issue of Helm trying to install the instance when the CRDs definitions are not yet properly loaded: this is the root cause of the issue tracked here.

We tried to install the CRD instance with a [helm _post-installation_ hook](https://helm.sh/docs/topics/charts_hooks/) (#439): so, the CRD instance will be installed after all the rest. Anyway, since there is no check from Helm that the CRDs definitions are ready in the Kubernetes API before running the post-installation hook, there is still a chance that the CRD instance installation is attempted when the CRDs definitions are not ready yet.

The solution provided by Helm (v3.8.0+) is to [move the CRDs definition in a dedicated folder]( https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you) (i.e., the `crds` folder). While it would solve this issue, would create new ones: CRDs would not be deleted nor updated with this method.
Moreover, testing this method with our operator, upgrading from the current stable failed, ending up with no Elemental CRDs definitions (my guess is that the "old" CRDs were detected from the new chart, so they were not installed/upgraded, but at the same time since the current stable chart has the CRDs definition installed as common resources, they were eventually deleted).

The other solution provided in the [Helm best practices](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-2-separate-charts) is to have separate charts for the CRDs definition and for all the rest.

One other option we could consider is to have the main chart as usual but a separate chart with the CRDs instances (which right now would be just the `ManagedOSVersionChannel` resource). This way, consumers installing the operator from the command line, could opt out from the default configuration or install it after installing the Elemental Operator chart.
Some investigation on what would happen (and how to manage) deleting the Elemental Operator chart while the resource chart is still there is required.

**Comment by kkaempf** [comment]
> [Helm best practices](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-2-separate-charts) is to have separate charts
 
Then that's the way to go, no ? 🤔 

Given that we're targeting Rancher Marketplace for elemental-operator (and it's my understanding that Rancher marketplace supports dependencies), having CLI users run two `helm` calls is acceptable imho.

**Comment by fgiudici** [comment]
> > [Helm best practices](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-2-separate-charts) is to have separate charts
> 
> Then that's the way to go, no ? thinking
> 
> Given that we're targeting Rancher Marketplace for elemental-operator (and it's my understanding that Rancher marketplace supports dependencies), having CLI users run two `helm` calls is acceptable imho.

Yep, I agree. The solution is to go with two charts, and are either point 2. or point 3.  (basically two chart, in one you put crds in the other the crds instances as per Helm best practices... they just differ in which one you put all the other templated common resources).
We should check which among 2. and 3. gives us the less troubles on updating from the current chart and on deletion.

**Comment by davidcassany** [comment]
I think ideally we should go with the second option. However this might not a straight forward path in terms of upgrades and uninstalls, requires some testing. My take is that if it simply results on a simple dual execution for every operation (install one after the other, upgrade one after the other, uninstall one after the other in reverse order) this should be the way to go. This should be relatively simple to code it mostly has impact to tests and docs.

**Comment by agracey** [comment]
https://hypper.io/ ...

**Comment by fgiudici** [comment]
> Are you sure that it won't update CRDs? I'm pretty sure I've seen that done before but maybe I'm missing something

For the proposed solution 1 above  (a separate crds directory), [that's clearly stated in the helm docs](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations).
For the dual charts solutions, crds updates  would work 👍🏼

**Comment by davidcassany** [comment]
#448 is the implementation of solution 3. Solution 2 is a breaking change, we can't afford that without a major version bump to make explicit there is no backward compatibility.

**Comment by mattfarina** [comment]
Here are my 2 cents....

- The same chart putting CRDs into a cluster can have a problem also putting CRs into the cluster. This is because the time it takes for the CRDs to be fully registered to be used. The `crds` directory should work for this but we fixed a recent bug on the issue where there were issues with Helm picking up the newly registered CRD. I think that's fixed in the latest Helm but there's a different bug there that's due to an upstream k8s bug and Helm 3.12 has issues with some clusters. Once k8s releases a fix for their packages Helm will release a fix for that issue.
- Rancher charts, in general, use the pattern of having a chart just for CRDs and then a chart for everything else. This is the common pattern and you can see it in the charts [here](https://github.com/rancher/charts).

At what point can you make chart breaking changes? I'm not suggesting you do. I'm curious what stops you or when you can do it.

**Comment by davidcassany** [comment]
> Here are my 2 cents....
> 
>* The same chart putting CRDs into a cluster can have a problem also putting CRs into the cluster. This is because the time it takes for the CRDs to be fully registered to be used. 

Absolutely, this is what this issue is about. We face sporadic race issues between trying to create CRs before they are fully registered.

> The `crds` directory should work for this but we fixed a recent bug on the issue where there were issues with Helm picking up the newly registered CRD. 

Indeed the `crds` directory works for the former issue, however then the problem is maintaining CRDs with a different flow.  Up to now updating and uninstalling chart is also dealing with CRDs. So then we have to provide our custom logic to handle its life cycle. We don't need anything special, hence having the CRDs as part of a chart works for us, they get updated on changes and are deleted on uninstall.
 
>* Rancher charts, in general, use the pattern of having a chart just for CRDs and then a chart for everything else. This is the common pattern and you can see it in the charts [here](https://github.com/rancher/charts).
 
Yes this is what we are aiming for. However just noted that the move is not so smooth. So the ideal goal would be something like having these two charts:

* `elemental-operator-crds` (only including the CRDs)
* `elemental-operator` (former chart without the CRDs)

In that case upgrades are tricky as if someone upgrades the installed release with the new `elemental-operator`  chart that will actually erase the CRDs (as they are no longer there) and associated CRs loosing relevant data (nodes would have to be re-provisioned) and leaving the installation in an inconsistent state (the controller running without the CRDs it controls).

So that we were thinking on a slightly different approach. Having still to charts:

* `elemental-operator` (former chart with CRDs, deployment, etc. but no CRs)
* `elemental-operator-defaults`  (just default CRs and other eventual pre-loaded data we might consider in the future)

This way chart install/upgrade/uninstall works as it used to be and it is really unlikely to break the current deployment on upgrades my messing with the two charts.
 
> At what point can you make chart breaking changes? I'm not suggesting you do. I'm curious what stops you or when you can do it.

Indeed, that's a good question. The risk of breaking a current deployment on edge requiring manual intervention to be fixed (having to hard reset nodes) is our no go. When can we consider it? I'd say a new major bump (so we easily claim certain upgrade constraints) having a consolidated way of remote nodes reset, so we can define a smooth process to apply the change and we have good alternatives to _recover_ in case something fails or if the upgrade was not properly executed.

**Comment by mattfarina** [comment]
@davidcassany I want to give a big warning about uninstalling CRDs. It's useful to read the Helm HIP titled [CRD Handling in Helm 3](https://github.com/helm/community/blob/main/hips/hip-0011.md). There is a lot of detail on ways people (users/customers) can have problems.

Uninstalling CRDs will cause all the CRs based on those to be deleted as well. If a CRD chart is uninstalled unintentionally, it can lead to data loss. I have seen data loss from this type of situation.

This is why the CRD chart is separate and why the `crds` directory doesn't deal with uninstalls. It's common to uninstall and reinstall a chart to fix a problem.

In what ever setup you do, please consider the unhappy paths.

**Comment by davidcassany** [comment]
> @davidcassany I want to give a big warning about uninstalling CRDs. It's useful to read the Helm HIP titled [CRD Handling in Helm 3](https://github.com/helm/community/blob/main/hips/hip-0011.md). There is a lot of detail on ways people (users/customers) can have problems.

Yes thanks, I see you point. In fact this has been always the intention, however I was not seeing a path on how to transition from the single chart to the splitted charts approach. Finally I think I came up to a reasonable approach on #447  by using helm hooks to make the transition pretty smooth, however still need to run some further tests.
